File: 02-python/dive.py
instructions = [line.rstrip().split(" ") for line in open("input").readlines()]

# How to code abstractions in Python? I guess with classes?
# I have no idea I just have an exam tomorrow where I need to hack a bit of Python. 
# If you are thinking about hiring me this is a joke, my Python skills are "intermediate" (:

def solve_1(movements):
    pos = 0
    depth = 0
    for move in movements:
        step = int(move[1])
        if move[0] == "forward":
            pos += step
        elif move[0] == "down":
            depth += step
        elif move[0] == "up":
            depth -= step
        else:
            raise ValueError(f"Unrecognized direction name: {move}.")
    return pos*depth

print(solve_1(instructions))

def solve_2(movements):
    pos = 0
    depth = 0
    aim = 0
    for move in movements:
        step = int(move[1])
        if move[0] == "forward":
            pos += step
            depth += step*aim
        elif move[0] == "down":
            aim += step
        elif move[0] == "up":
            aim -= step
        else:
            raise ValueError(f"Unrecognized direction name: {move}.")
    return pos*depth

print(solve_2(instructions))            

File: 2-julia/dive.jl
abstract type Direction end

struct Forward <: Direction end
struct Up <: Direction end
struct Down <: Direction end

function direction(name)
    if name == "forward"
        return Forward
    elseif name == "up"
        return Up
    elseif name == "down"
        return Down
    else 
        @error "Unrecognized direction name"
    end
end 

struct Movement{T}
    n::Int
end 

movement(mov) = Movement{direction(mov[1])}(parse(Int, mov[2]))

const input = movement.(split.(readlines("input"), " "))

mutable struct Pos
    x::Int
    y::Int
end 

Pos() = Pos(0, 0)

apply!(pos::Pos, mov::Movement{Forward}) = pos.x += mov.n
apply!(pos::Pos, mov::Movement{Up})      = pos.y -= mov.n
apply!(pos::Pos, mov::Movement{Down})    = pos.y += mov.n

function solve(input, pos)
    map(mov -> apply!(pos, mov), input)
    return pos.x * pos.y
end 

answ_1 = solve(input, Pos())

# Part 2
mutable struct PosAim
    x::Int
    y::Int
    aim::Int
end 

PosAim() = PosAim(0, 0, 0)

function apply!(pos::PosAim, mov::Movement{Forward})
    pos.x += mov.n
    pos.y += mov.n * pos.aim
    return nothing
end 
apply!(pos::PosAim, mov::Movement{Up})   = pos.aim -= mov.n
apply!(pos::PosAim, mov::Movement{Down}) = pos.aim += mov.n

answ_2 = solve(input, PosAim())
